app/lib/client_sorter.rb

```diff
@@ -32,22 +32,52 @@ def filtered_and_sorted_clients(default_order: nil)
 
   def filtered_clients
     clients = if current_user&.greeter?
-                # Greeters should only have "search" access to clients in intake stage AND clients assigned to them.
-                @clients.greetable || Client.joins(:tax_returns).where(tax_returns: { assigned_user: current_user }).distinct
+                @clients
               else
                 @clients.after_consent
               end
     # Force an inner join to `intakes` to exclude clients from previous years
-    clients = clients.distinct.joins(:intake)
+    clients = clients.joins(:intake)
     clients = clients.where(intake: Intake.where(type: "Intake::CtcIntake")) if @filters[:ctc_client].present?
-    clients = clients.where(tax_returns: { current_state: TaxReturnStateMachine::STATES_BY_STAGE[@filters[:stage]] }) if @filters[:stage].present?
     clients = clients.where.not(flagged_at: nil) if @filters[:flagged].present?
-    clients = clients.where(tax_returns: { assigned_user: limited_user_ids }) unless limited_user_ids.empty?
-    clients = clients.where(tax_returns: { year: @filters[:year] }) if @filters[:year].present?
-    clients = clients.where(tax_returns: { current_state: @filters[:status] }) if @filters[:status].present?
-    clients = clients.where.not(tax_returns: { current_state: TaxReturnStateMachine::EXCLUDED_FROM_SLA }) if @filters[:active_returns].present?
+
+    tax_return_filters = {}
+    if @filters[:stage].present?
+      tax_return_filters[:stage] = @filters[:stage]
+    end
+    if @filters[:year].present?
+      tax_return_filters[:year] = @filters[:year].to_i
+    end
+    if @filters[:status].present?
+      tax_return_filters[:current_state] = @filters[:status]
+    end
+    if @filters[:active_returns].present?
+      tax_return_filters[:active] = @filters[:active_returns].in?([true, "true"])
+    end
+    if @filters[:service_type].present?
+      tax_return_filters[:service_type] = @filters[:service_type]
+    end
+
+    tax_return_filters_expanded = []
+
+    if limited_user_ids.present?
+      special_tax_return_filters = current_user&.greeter? ? tax_return_filters.merge(greetable: true) : tax_return_filters
+      limited_user_ids.each do |id|
+        tax_return_filters_expanded << special_tax_return_filters.merge(assigned_user_id: id)
+      end
+    elsif current_user&.greeter?
+      tax_return_filters_expanded = [tax_return_filters.merge(greetable: true), tax_return_filters.merge(assigned_user_id: current_user.id)]
+    else
+      tax_return_filters_expanded = [tax_return_filters]
+    end
+
+    clients = tax_return_filters_expanded.map do |tax_return_filters|
+      clients.where("filterable_tax_return_properties @> '[#{tax_return_filters.to_json}]'::jsonb")
```

Comment:
I think we can use ActiveRecord `?` here safely and effectively. Try instead:

```
clients.where("filterable_tax_return_properties @> [?]::jsonb", tax_return_filters.to_json)
```

I think this works OK based on the following testing:

rails c

```
irb(main):014:0> puts AdminRole.where('primary_first_name @> [?]::jsonb', {hi: "there's"}.to_json).select('
primary_first_name').to_sql
SELECT primary_first_name FROM "admin_roles" WHERE (primary_first_name @> ['{"hi":"there''s"}']::jsonb)
=> nil

```

psql

```
vita-min_development=# select '{"hi":"there''s"}'::jsonb as ok from intakes limit 1;
        ok         
-------------------
 {"hi": "there's"}
(1 row)

```

It looks like `''''` is a single apostrophe in SQL:

```
vita-min_development=# select '''' as ok from intakes limit 1;
 ok 
----
 '
(1 row)

```

It looks like without the `?`, one does run into problems.

```
irb(main):016:0> val = {hi: "there's"} ; AdminRole.where("primary_first_name @> '[#{val.to_json}]'::jsonb")

  AdminRole Load (2.8ms)  SELECT "admin_roles"."id", "admin_roles"."created_at", "admin_roles"."updated_at" FROM "admin_roles" WHERE (primary_first_name @> '[{"hi":"there's"}]'::jsonb) /* loading for inspect */ LIMIT $1  [["LIMIT", 11]]
Traceback (most recent call last):
ActiveRecord::StatementInvalid (PG::SyntaxError: ERROR:  syntax error at or near "s")
LINE 1: ...oles" WHERE (primary_first_name @> '[{"hi":"there's"}]'::jso...
                                                             ^
```

Happy to Tuple or not. Let me know what you think.

Reply:
Rather

```
clients.where("filterable_tax_return_properties @> '[?]'::jsonb", tax_return_filters.to_json)
```

Reply:
Hmm I might be wrong about where it's OK to put these `'`s